File: sdks/node/src/seed/discovery/mdns.ts
/**
 * `MdnsDiscovery` — opt-in mDNS provider (ADR-0016a §D6 Phase 1.5).
 *
 * Imports `multicast-dns` via the seed's service type
 * `_cognitum._tcp.local` and returns the TXT records emitted by
 * `seed/src/cognitum-agent/src/discovery.rs:137-180` as
 * {@link DiscoveredPeer}s.
 *
 * The wire library is declared as a **peerDependency** so the core
 * `@cognitum/sdk` install stays lean — callers who want mDNS opt into
 * the dep by importing this file via the subpath
 * `@cognitum/sdk/seed/discovery/mdns`.
 *
 * Phase 3 punted items (tracked in `docs/adr/0015c-...` §"Phase 3 —
 * mDNS discovery"):
 *
 * - PTR → SRV → A/AAAA chain lookup. Today we trust the TXT-record host
 *   hint plus a fallback to the seed's default `.local` hostname.
 *
 * `fp=` (cert fingerprint) pinning — ADR-0015c Phase 3 §fp= cert
 * pinning (2026-04-23) — is now parsed into
 * {@link DiscoveredPeer.tlsFingerprint} and wired into the per-peer
 * TLS handshake by `src/seed/transport.ts`. A mismatch throws
 * {@link TlsPinError}; see that class' docstring for semantics.
 */

import type { DiscoveryProvider, DiscoveredPeer } from "./types.js";
import { ConfigError } from "../../errors.js";

// Node built-ins used below. Declared rather than imported so the
// `tsup --dts` pass — which runs without `@types/node` in lib — still
// emits types cleanly. Mirrors the pattern used across `src/seed/*.ts`.
declare const TextDecoder: {
  new (encoding?: string): { decode(input?: Uint8Array): string };
};
declare function setTimeout(cb: () => void, ms?: number): unknown;

/**
 * Structural shim for the `multicast-dns` module. Declared here so the
 * SDK can compile without the dep installed — `peerDependencies`
 * install the real module at the caller's discretion.
 */
interface MdnsInstance {
  query(
    name: string | { questions: Array<{ name: string; type: string }> },
    type?: string,
  ): void;
  on(event: "response", cb: (packet: MdnsPacket) => void): void;
  destroy(cb?: () => void): void;
}

interface MdnsPacket {
  answers?: MdnsAnswer[];
  additionals?: MdnsAnswer[];
}

interface MdnsAnswer {
  name: string;
  type: string;
  data: unknown;
}

/** Factory signature exported by `multicast-dns`. */
type MdnsFactory = (opts?: {
  multicast?: boolean;
  interface?: string;
  port?: number;
  loopback?: boolean;
}) => MdnsInstance;

/** Options for {@link MdnsDiscovery}. */
export interface MdnsDiscoveryOptions {
  /**
   * DNS-SD service type. Defaults to the seed's advertised
   * `_cognitum._tcp.local` per `seed/src/cognitum-agent/src/discovery.rs:99`.
   */
  serviceType?: string;
  /**
   * Collection window in ms. Responses arriving after this fire-and-
   * forget window are dropped. Default 500ms — long enough to catch
   * seeds on the same LAN, short enough that `client.rediscover()` is
   * snappy. Must be a positive number.
   */
  timeoutMs?: number;
  /**
   * Default TCP port to construct URLs with when the TXT record omits
   * `port=`. Defaults to 8443 (seed HTTPS).
   */
  defaultPort?: number;
  /**
   * TLS scheme for constructed URLs. `https` by default; set to `http`
   * only for mesh test harnesses.
   */
  scheme?: "https" | "http";
  /**
   * Inject the `multicast-dns` factory (tests + advanced callers). When
   * omitted, {@link MdnsDiscovery.discover} dynamically imports
   * `multicast-dns` the first time it is invoked. The dynamic import
   * keeps the dep optional at install time.
   */
  mdnsFactory?: MdnsFactory;
}

const DEFAULT_SERVICE_TYPE = "_cognitum._tcp.local";
const DEFAULT_TIMEOUT_MS = 500;
const DEFAULT_PORT = 8443;

/**
 * Opt-in mDNS discovery provider. Returns peers whose TXT records
 * answer a PTR query for the configured service type.
 *
 * @example
 * ```ts
 * import { SeedClient } from "@cognitum/sdk/seed";
 * import { MdnsDiscovery } from "@cognitum/sdk/seed/discovery/mdns";
 *
 * const client = new SeedClient({
 *   endpoints: new MdnsDiscovery(),
 *   tls: { insecure: true }, // dev only
 * });
 * ```
 */
export class MdnsDiscovery implements DiscoveryProvider {
  private readonly serviceType: string;
  private readonly timeoutMs: number;
  private readonly defaultPort: number;
  private readonly scheme: "https" | "http";
  private readonly mdnsFactory: MdnsFactory | undefined;
  /** Lazily instantiated mDNS wire handle — reused across `discover()` calls. */
  private instance: MdnsInstance | undefined;

  constructor(opts: MdnsDiscoveryOptions = {}) {
    this.serviceType = opts.serviceType ?? DEFAULT_SERVICE_TYPE;
    const timeout = opts.timeoutMs ?? DEFAULT_TIMEOUT_MS;
    if (!Number.isFinite(timeout) || timeout <= 0) {
      throw new ConfigError(
        `MdnsDiscovery.timeoutMs must be a positive number (got ${timeout})`,
      );
    }
    this.timeoutMs = timeout;
    this.defaultPort = opts.defaultPort ?? DEFAULT_PORT;
    this.scheme = opts.scheme ?? "https";
    this.mdnsFactory = opts.mdnsFactory;
  }

  /**
   * Convenience constructor matching the ADR example surface —
   * `MdnsDiscovery.default()` reads as "use the seed's published
   * defaults" at call-sites.
   */
  static default(): MdnsDiscovery {
    return new MdnsDiscovery();
  }

  async discover(): Promise<DiscoveredPeer[]> {
    const mdns = await this.ensureInstance();
    const seen = new Map<string, DiscoveredPeer>();

    const collector = (packet: MdnsPacket): void => {
      const records = [
        ...(packet.answers ?? []),
        ...(packet.additionals ?? []),
      ];
      for (const rec of records) {
        if (rec.type !== "TXT") continue;
        if (!this.matchesService(rec.name)) continue;
        const parsed = parseTxtRecord(rec.data);
        if (!parsed) continue;
        const peer = this.peerFromTxt(parsed, rec.name);
        if (peer && !seen.has(peer.url)) {
          seen.set(peer.url, peer);
        }
      }
    };

    mdns.on("response", collector);
    mdns.query({
      questions: [{ name: this.serviceType, type: "PTR" }],
    });

    await sleep(this.timeoutMs);

    // `multicast-dns` lacks a typed `off`; best-effort detach to keep
    // the listener list small across repeated `discover()` calls.
    const off = (mdns as unknown as {
      removeListener?: (e: string, cb: (p: MdnsPacket) => void) => void;
    }).removeListener;
    off?.call(mdns, "response", collector);

    return Array.from(seen.values());
  }

  async close(): Promise<void> {
    const inst = this.instance;
    this.instance = undefined;
    if (!inst) return;
    await new Promise<void>((resolve) => {
      try {
        inst.destroy(() => resolve());
      } catch {
        resolve();
      }
    });
  }

  // ------------------------------------------------------------------ //
  // internals                                                           //
  // ------------------------------------------------------------------ //

  private async ensureInstance(): Promise<MdnsInstance> {
    if (this.instance) return this.instance;
    const factory = this.mdnsFactory ?? (await loadMdnsFactory());
    this.instance = factory();
    return this.instance;
  }

  private matchesService(recordName: string): boolean {
    // TXT records may be attached to either `_cognitum._tcp.local` or
    // a per-instance name like `cognitum-61bc._cognitum._tcp.local`.
    return (
      recordName === this.serviceType ||
      recordName.endsWith(`.${this.serviceType}`)
    );
  }

  private peerFromTxt(
    txt: Record<string, string>,
    recordName: string,
  ): DiscoveredPeer | undefined {
    const port =
      Number.parseInt(txt.port ?? "", 10) || this.defaultPort;
    const host = hostFromRecordName(recordName) ?? txt.host;
    if (!host) return undefined;
    const url = `${this.scheme}://${host}:${port}`;
    const tlsFingerprint = parseFingerprint(txt.fp);
    const peer: DiscoveredPeer = {
      url,
      deviceId: txt.id,
    };
    if (tlsFingerprint !== undefined) {
      peer.tlsFingerprint = tlsFingerprint;
    }
    return peer;
  }
}

/**
 * Parse a TXT-record `fp=` value into the canonical hex-only form used
 * by the transport layer's pinning check.
 *
 * Accepts:
 *  - `sha256:<hex>` (the form documented in ADR-040 FINDING-28)
 *  - bare `<hex>` (the form the seed emits today, per
 *    `seed/src/cognitum-agent/src/discovery.rs:162`)
 *
 * Normalisation: strip any `sha256:` prefix (case-insensitive), strip
 * colons (some DNS-SD responders format fingerprints as
 * `aa:bb:cc:...`), lowercase. A string that contains any non-hex
 * character after normalisation is rejected — returning `undefined`
 * rather than throwing so a malformed TXT doesn't tank the whole
 * `discover()` batch.
 *
 * Exported for unit testing (`tests/seed/unit/discovery-mdns-fp.test.ts`).
 */
export function parseFingerprint(raw: string | undefined): string | undefined {
  if (raw === undefined || raw === null) return undefined;
  if (typeof raw !== "string") return undefined;
  let s = raw.trim();
  if (s.length === 0) return undefined;
  // Strip `sha256:` / `SHA256:` prefix if present.
  const colonIdx = s.indexOf(":");
  if (colonIdx > 0 && colonIdx <= 7) {
    const prefix = s.slice(0, colonIdx).toLowerCase();
    if (prefix === "sha256" || prefix === "sha-256") {
      s = s.slice(colonIdx + 1);
    }
  }
  // Strip any remaining colons (legacy `aa:bb:cc` style).
  s = s.replace(/:/g, "").toLowerCase();
  if (s.length === 0) return undefined;
  if (s.length % 2 !== 0) return undefined;
  if (!/^[0-9a-f]+$/.test(s)) return undefined;
  return s;
}

/**
 * Parse a TXT record's `data` field into a key=value map. `multicast-dns`
 * exposes the data as `Buffer | Buffer[]` (one Buffer per entry).
 */
function parseTxtRecord(data: unknown): Record<string, string> | undefined {
  const entries: unknown[] = Array.isArray(data) ? data : [data];
  const out: Record<string, string> = {};
  let any = false;
  const decoder = new TextDecoder("utf-8");
  for (const e of entries) {
    let s: string | undefined;
    if (typeof e === "string") {
      s = e;
    } else if (e instanceof Uint8Array) {
      s = decoder.decode(e);
    } else if (
      e !== null &&
      typeof e === "object" &&
      "toString" in e &&
      typeof (e as { toString: unknown }).toString === "function"
    ) {
      s = String(e);
    }
    if (!s) continue;
    const idx = s.indexOf("=");
    if (idx <= 0) continue;
    const key = s.slice(0, idx).trim().toLowerCase();
    const val = s.slice(idx + 1);
    if (!key) continue;
    out[key] = val;
    any = true;
  }
  return any ? out : undefined;
}

/**
 * Extract the instance host from a fully-qualified DNS-SD record name
 * like `cognitum-61bc._cognitum._tcp.local` → `cognitum-61bc.local`.
 * Returns `undefined` when the record name is the bare service type.
 */
function hostFromRecordName(recordName: string): string | undefined {
  // Strip the `._cognitum._tcp` middle labels and re-append `.local`.
  const m = recordName.match(/^([^.]+)\._[^.]+\._tcp\.(.+)$/);
  if (!m) return undefined;
  return `${m[1]}.${m[2]}`;
}

function sleep(ms: number): Promise<void> {
  return new Promise((r) => setTimeout(r, ms));
}

/**
 * Dynamically import `multicast-dns`. Surfaces a typed {@link ConfigError}
 * when the dep isn't installed so the failure mode is clearly a
 * missing-peer-dep rather than an opaque module resolution error.
 */
async function loadMdnsFactory(): Promise<MdnsFactory> {
  try {
    // Indirected through a string var so tsc doesn't try to resolve the
    // optional peer dep at type-check time. `multicast-dns` is declared
    // as a peerDependency — real callers who opt in will install it.
    const spec = "multicast-dns";
    const mod = (await import(spec)) as unknown as
      | { default: MdnsFactory }
      | MdnsFactory;
    const factory = (mod as { default?: MdnsFactory }).default ?? (mod as MdnsFactory);
    if (typeof factory !== "function") {
      throw new Error("multicast-dns module did not export a factory");
    }
    return factory;
  } catch (err) {
    throw new ConfigError(
      `mDNS discovery requires the optional 'multicast-dns' dependency. ` +
        `Install it: npm install multicast-dns (original error: ${
          err instanceof Error ? err.message : String(err)
        })`,
    );
  }
}

File: sdks/rust/src/seed/discovery/mdns.rs
//! mDNS / DNS-SD discovery (ADR-0016a §D6, Phase 3).
//!
//! Gated behind `feature = "mdns"`. Performs a single DNS-SD browse for
//! the configured service type (default `_cognitum._tcp.local.`) and
//! parses the TXT record emitted by `seed/src/cognitum-agent/src/discovery.rs`
//! into a set of [`DiscoveredPeer`]s.
//!
//! `mdns-sd` has no async runtime dependency — it runs a blocking
//! background thread and delivers events through a `flume` receiver. We
//! drive it from a `tokio::task::spawn_blocking` so it composes cleanly
//! with the SDK's `tokio` async runtime.

use std::collections::HashMap;
use std::time::Duration;

use super::{DiscoveredPeer, Discovery};
use crate::error::Error;

/// Default DNS-SD service type advertised by `cognitum-agent` (see
/// `seed/src/cognitum-agent/src/discovery.rs:3-5`).
pub const DEFAULT_SERVICE_TYPE: &str = "_cognitum._tcp.local.";

/// Default maximum time to accumulate mDNS responses before returning.
///
/// Most LAN mDNS responders answer in <100 ms; 2 s is the same budget
/// `avahi-browse -t` uses by default. Tune via
/// [`MdnsDiscovery::browse_duration`] for CI or slow links.
pub const DEFAULT_BROWSE_DURATION: Duration = Duration::from_secs(2);

/// Default port used when a TXT record is missing the `port` key. Matches
/// the cognitum-agent default HTTPS listener.
pub const DEFAULT_PORT: u16 = 8443;

/// One-shot mDNS discovery provider (ADR-0016a §D6).
///
/// Cheap to clone — holds only configuration. Each `discover()` call
/// spins up a fresh `ServiceDaemon` so failures are self-contained and
/// there's no background traffic between calls.
#[derive(Debug, Clone)]
pub struct MdnsDiscovery {
    service_type: String,
    browse_duration: Duration,
    scheme: &'static str,
    default_port: u16,
}

impl Default for MdnsDiscovery {
    fn default() -> Self {
        Self {
            service_type: DEFAULT_SERVICE_TYPE.to_owned(),
            browse_duration: DEFAULT_BROWSE_DURATION,
            scheme: "https",
            default_port: DEFAULT_PORT,
        }
    }
}

impl MdnsDiscovery {
    /// Construct with defaults (`_cognitum._tcp.local.`, 2 s browse, https).
    pub fn new() -> Self {
        Self::default()
    }

    /// Start a fluent builder.
    pub fn builder() -> MdnsDiscoveryBuilder {
        MdnsDiscoveryBuilder::default()
    }

    /// Currently configured service type.
    pub fn service_type(&self) -> &str {
        &self.service_type
    }

    /// Currently configured browse duration.
    pub fn browse_duration(&self) -> Duration {
        self.browse_duration
    }
}

/// Fluent builder for [`MdnsDiscovery`]. Use [`MdnsDiscovery::builder`]
/// to construct one.
#[derive(Debug, Clone)]
pub struct MdnsDiscoveryBuilder {
    service_type: String,
    browse_duration: Duration,
    scheme: &'static str,
    default_port: u16,
}

impl Default for MdnsDiscoveryBuilder {
    fn default() -> Self {
        let defaults = MdnsDiscovery::default();
        Self {
            service_type: defaults.service_type,
            browse_duration: defaults.browse_duration,
            scheme: defaults.scheme,
            default_port: defaults.default_port,
        }
    }
}

impl MdnsDiscoveryBuilder {
    /// Override the DNS-SD service type. Must end in `.local.` to match
    /// link-local semantics; trailing dot is required by `mdns-sd`.
    pub fn service_type(mut self, st: impl Into<String>) -> Self {
        self.service_type = st.into();
        self
    }

    /// Maximum wall time to wait for responses. Defaults to 2 seconds.
    pub fn browse_duration(mut self, d: Duration) -> Self {
        self.browse_duration = d;
        self
    }

    /// URL scheme to construct for discovered peers. The seed HTTPS
    /// listener is the default; override to `"http"` only for local
    /// testing against a plaintext mock.
    pub fn scheme(mut self, scheme: &'static str) -> Self {
        self.scheme = scheme;
        self
    }

    /// Port to use when the TXT record does not supply a `port=` entry.
    pub fn default_port(mut self, port: u16) -> Self {
        self.default_port = port;
        self
    }

    /// Finalise.
    pub fn build(self) -> MdnsDiscovery {
        MdnsDiscovery {
            service_type: self.service_type,
            browse_duration: self.browse_duration,
            scheme: self.scheme,
            default_port: self.default_port,
        }
    }
}

#[async_trait::async_trait]
impl Discovery for MdnsDiscovery {
    async fn discover(&self) -> Result<Vec<DiscoveredPeer>, Error> {
        let service_type = self.service_type.clone();
        let budget = self.browse_duration;
        let scheme = self.scheme;
        let default_port = self.default_port;

        // `mdns-sd`'s `ServiceDaemon` holds sockets and spawns its own
        // thread — wrap the whole browse in `spawn_blocking` so we don't
        // hog a tokio worker and so that socket setup errors don't
        // propagate as async panics.
        let handle = tokio::task::spawn_blocking(move || -> Result<Vec<DiscoveredPeer>, Error> {
            let daemon = mdns_sd::ServiceDaemon::new().map_err(|e| Error::Api {
                code: 0,
                message: format!("mdns: failed to start ServiceDaemon: {e}"),
            })?;
            let receiver = daemon.browse(&service_type).map_err(|e| Error::Api {
                code: 0,
                message: format!("mdns: browse failed for `{service_type}`: {e}"),
            })?;

            // Deduplicate on `instance_name` — DNS-SD sends both
            // `ServiceFound` and later `ServiceResolved`; we only care
            // about the resolved form.
            let mut seen: HashMap<String, DiscoveredPeer> = HashMap::new();
            let started = std::time::Instant::now();

            loop {
                let elapsed = started.elapsed();
                if elapsed >= budget {
                    break;
                }
                let remaining = budget - elapsed;
                match receiver.recv_timeout(remaining) {
                    Ok(mdns_sd::ServiceEvent::ServiceResolved(info)) => {
                        let latency = elapsed.as_millis().min(u128::from(u32::MAX)) as u32;
                        let fullname = info.get_fullname().to_owned();
                        let peer = resolve_info_to_peer(&info, scheme, default_port, latency);
                        if let Some(p) = peer {
                            seen.insert(fullname, p);
                        }
                    }
                    // Ignore `ServiceFound` / `SearchStarted` / etc —
                    // we want the resolved form with TXT keys.
                    Ok(_) => {}
                    // `flume::RecvTimeoutError` isn't re-exported by
                    // mdns-sd; either variant ends the browse loop, so
                    // a bare catch-all is semantically equivalent.
                    Err(_) => break,
                }
            }

            // Best-effort shutdown — the daemon drops its socket when the
            // handle goes out of scope, but call the explicit API so
            // tests don't leak ports across runs.
            let _ = daemon.shutdown();

            let mut out: Vec<DiscoveredPeer> = seen.into_values().collect();
            // Stable order: sort by URL so rediscovery produces a
            // deterministic `list_index` for unchanged networks.
            out.sort_by(|a, b| a.url.cmp(&b.url));
            Ok(out)
        })
        .await
        .map_err(|e| Error::Api {
            code: 0,
            message: format!("mdns: discover task panicked: {e}"),
        })?;

        handle
    }
}

/// Turn a `mdns_sd::ResolvedService` into a [`DiscoveredPeer`].
///
/// Returns `None` when the advertised record has no address — we cannot
/// build a URL without a host, and the alternative (guessing localhost)
/// would silently mis-route writes.
fn resolve_info_to_peer(
    info: &mdns_sd::ResolvedService,
    scheme: &str,
    default_port: u16,
    observed_ms: u32,
) -> Option<DiscoveredPeer> {
    // Prefer a numeric IPv4; fall back to any other address; then to the
    // hostname. ResolvedService.addresses is a HashSet<ScopedIp> and
    // ScopedIp Display-prints to its textual form.
    let mut ipv4: Option<String> = None;
    let mut other: Option<String> = None;
    for a in info.get_addresses() {
        let s = a.to_string();
        // ScopedIp's Display includes "%scope" for link-local; the
        // ScopedIpV4 case is always IPv4.
        if s.contains('.') && !s.contains(':') {
            ipv4 = Some(s);
            break;
        }
        other.get_or_insert(s);
    }
    let host: String = ipv4
        .or(other)
        .unwrap_or_else(|| info.get_hostname().trim_end_matches('.').to_owned());
    if host.is_empty() {
        return None;
    }

    // Port: prefer the TXT `port=` key if present (seed sets it), else
    // fall back to the DNS-SD record port, then the default.
    let txt_port = info
        .get_property_val_str("port")
        .and_then(|s| s.parse::<u16>().ok());
    let record_port = info.get_port();
    let port = txt_port
        .or(if record_port == 0 {
            None
        } else {
            Some(record_port)
        })
        .unwrap_or(default_port);

    // IPv6 hosts need bracketing in a URL; IPv4 and hostnames don't.
    let host_literal = if host.contains(':') && !host.starts_with('[') {
        format!("[{host}]")
    } else {
        host
    };

    let url = format!("{scheme}://{host_literal}:{port}");
    let device_id = info.get_property_val_str("id").map(str::to_owned);
    // `fp=sha256:<hex>` per seed/src/cognitum-agent/src/discovery.rs.
    // Normalise to lowercase hex without colons so the FingerprintPinVerifier
    // comparison is trivial.
    let fp = info
        .get_property_val_str("fp")
        .map(super::normalize_fingerprint);

    Some(
        DiscoveredPeer::new(url)
            .with_latency_ms(observed_ms)
            .set_device_id(device_id)
            .set_tls_fingerprint(fp),
    )
}

// Small helpers: fluent `set_device_id(Option<String>)` /
// `set_tls_fingerprint(Option<String>)` — the existing `with_*`
// constructors take owned values, so these keep the call site clean
// when the parsed TXT key may be missing.
impl DiscoveredPeer {
    fn set_device_id(mut self, id: Option<String>) -> Self {
        self.device_id = id;
        self
    }

    fn set_tls_fingerprint(mut self, fp: Option<String>) -> Self {
        self.tls_fingerprint = fp;
        self
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn builder_defaults_match_adr() {
        let d = MdnsDiscovery::builder().build();
        assert_eq!(d.service_type(), DEFAULT_SERVICE_TYPE);
        assert_eq!(d.browse_duration(), DEFAULT_BROWSE_DURATION);
    }

    #[test]
    fn builder_overrides_are_applied() {
        let d = MdnsDiscovery::builder()
            .service_type("_cognitum-dev._tcp.local.")
            .browse_duration(Duration::from_millis(500))
            .scheme("http")
            .default_port(18080)
            .build();
        assert_eq!(d.service_type(), "_cognitum-dev._tcp.local.");
        assert_eq!(d.browse_duration(), Duration::from_millis(500));
    }

    #[tokio::test]
    async fn discover_on_empty_network_returns_empty_fast() {
        // Smoke test: on CI / sandboxed containers multicast is usually
        // blocked. The browse budget elapses with no `ServiceResolved`
        // events and we return `Ok(vec![])`. Cap the budget tight so
        // the test doesn't pad suite runtime.
        let d = MdnsDiscovery::builder()
            .service_type("_cognitum-nope._tcp.local.")
            .browse_duration(Duration::from_millis(150))
            .build();
        let peers = d.discover().await.unwrap_or_default();
        assert!(peers.is_empty());
    }

    #[test]
    fn fingerprint_normalisation_strips_prefix_and_colons() {
        use super::super::normalize_fingerprint;
        // Seed wire form is `sha256:AA:BB:...` — we want lowercase hex.
        let got = normalize_fingerprint("sha256:AA:BB:CC:DD");
        assert_eq!(got, "aabbccdd");
        let got2 = normalize_fingerprint("  SHA256:AaBb  ");
        assert_eq!(got2, "aabb");
        // No prefix — still works.
        let got3 = normalize_fingerprint("AABB");
        assert_eq!(got3, "aabb");
    }

    #[test]
    fn set_tls_fingerprint_round_trip_on_discovered_peer() {
        let p = DiscoveredPeer::new("https://seed:8443").with_tls_fingerprint("sha256:AA:BB");
        assert_eq!(p.tls_fingerprint.as_deref(), Some("aabb"));
    }
}
